From 28125816a6e96fa046c46a90d32f55cbfec341df Mon Sep 17 00:00:00 2001
From: Aniket Randive <arandive@qti.qualcomm.com>
Date: Mon, 8 Sep 2025 15:41:39 +0530
Subject: [PATCH 04/12] i2c: i2c-qcom-geni: Add APIs to support CCU QUP client

Add follwing new exported APIs to support CCU QUP client,
- Bus recovery feature: Support recovery of IOS lines in both FIFO
and GSI modes during data transfers.
- Configure serial engine mode: Allow client to configure serial
engine mode in CDM or sync trigger mode, and set GPII channel
type to sync ch0, sync ch1, or regular channel.
- GPII reset feature: Enable client to initiate reset of any
particular GPII or all GPIIs of SE.

Apart from this added support for priority data transfer on
regular channel if priority bit is set.

Change-Id: Id8c7c7b18a8f9d631f3b208346d6f126aded2116

Signed-off-by: Aniket Randive <arandive@qti.qualcomm.com>
---
 drivers/i2c/busses/i2c-qcom-geni.c | 458 +++++++++++++++++++++++++++--
 1 file changed, 441 insertions(+), 17 deletions(-)

diff --git a/drivers/i2c/busses/i2c-qcom-geni.c b/drivers/i2c/busses/i2c-qcom-geni.c
index f79428c1abd7..d7e8c842eda9 100644
--- a/drivers/i2c/busses/i2c-qcom-geni.c
+++ b/drivers/i2c/busses/i2c-qcom-geni.c
@@ -102,16 +102,16 @@ struct geni_i2c_desc {
 
 /* CCU FRAME */
 #define CCU_DATA_OFFSET		8
-#define CCU_GPII_MASK		GENMASK(3, 0)
-#define CCU_ARB_LOST_MASK	GENMASK(16, 13)
-#define CCU_ARB_LOST_SHFT	13
+#define CCU_GPII_MASK		GENMASK(4, 0)
+#define CCU_ARB_LOST_MASK	GENMASK(15, 12)
+#define CCU_ARB_LOST_SHFT	12
 #define CCU_PKT_LEN_MASK	GENMASK(21, 16)
 #define CCU_PKT_LEN_SHFT	16
 #define CCU_FRAME_COOKIE_MASK	GENMASK(63, 32)
 #define CCU_FRAME_COOKIE_SHFT	32
 
 #define CCU_MAX_GPII		3
-#define REGULAR_GPII		2
+#define REGULAR_GPII		0
 #define MAX_NUM_MSGS		4096
 #define CCU_GET_GPII(val)	(val & CCU_GPII_MASK)
 #define CCU_GET_SYNC(val)	(val & BIT(5))
@@ -119,8 +119,8 @@ struct geni_i2c_desc {
 #define CCU_GET_NOP_FRAME(val)	(val & BIT(7))
 #define CCU_GET_INSERT_DELAY_BEFORE(val)	(val & BIT(8))
 #define CCU_GET_INSERT_DELAY_AFTER(val)		(val & BIT(9))
-#define CCU_GET_CH_LAST(val)		(val & BIT(11))
-#define CCU_GET_LAST(val)		(val & BIT(12))
+#define CCU_GET_CH_LAST(val)		(val & BIT(10))
+#define CCU_GET_LAST(val)		(val & BIT(11))
 #define CCU_GET_ARB_LOST_CNT(val)	((val & CCU_ARB_LOST_MASK) >> CCU_ARB_LOST_SHFT)
 #define CCU_GET_PKT_LEN(val)		((val & CCU_PKT_LEN_MASK) >> CCU_PKT_LEN_SHFT)
 #define CCU_GET_FRAME_COOKIE(val)	((val & CCU_FRAME_COOKIE_MASK) >> CCU_FRAME_COOKIE_SHFT)
@@ -204,6 +204,7 @@ struct geni_i2c_dev {
 	enum geni_se_xfer_mode xfer_mode;
 	bool is_ccu_qup;
 	struct ccu_i2c_config ccu;
+	int msgs;
 };
 
 struct geni_i2c_err_log {
@@ -252,6 +253,10 @@ static const struct geni_i2c_clk_fld geni_i2c_clk_map[] = {
 	{KHZ(1000), 1, 3,  9, 18},
 };
 
+static int setup_gpi_dma(struct geni_i2c_dev *gi2c);
+static void release_gpi_dma(struct geni_i2c_dev *gi2c);
+static void i2c_gpi_cb_result(void *cb, const struct dmaengine_result *result);
+
 static int log_level = -1;
 static ssize_t log_level_show(struct device *dev,
 			      struct device_attribute *attr, char *buf)
@@ -303,6 +308,375 @@ static void qcom_geni_i2c_conf(struct geni_i2c_dev *gi2c)
 	writel_relaxed(val, gi2c->se.base + SE_I2C_SCL_COUNTERS);
 }
 
+/*
+ * geni_i2c_power_on_off - This function is used to turn on or off all the clock
+ *                         and it will free the GPII resources in suspend path and
+ *                         allocate the resources in resume path.
+ * @adap: i2c adapter handle
+ * @suspend: true in case of suspend or false
+ *
+ * Return 0 on success or error on failure
+ */
+int geni_i2c_power_on_off(struct i2c_adapter *adap, bool suspend)
+{
+	struct geni_i2c_dev *gi2c;
+	int ret, i;
+
+	if (!adap) {
+		pr_err("%s: Adap is NULL\n", __func__);
+		return -EINVAL;
+	} else {
+		gi2c = adap->dev.driver_data;
+		if (!gi2c) {
+			pr_err("%s: I2C is NULL\n", __func__);
+			return -EINVAL;
+		}
+	}
+	if (suspend) {
+		for (i = 0; i < CCU_MAX_GPII; i++) {
+			if (gi2c->tx_c[i])
+				dmaengine_free_chan_resources(gi2c->tx_c[i]);
+
+			if (gi2c->rx_c[i])
+				dmaengine_free_chan_resources(gi2c->rx_c[i]);
+		}
+
+		/* Decrement the device usage counter and call pm runtime idle */
+		ret = pm_runtime_put_sync(gi2c->se.dev);
+		if (ret < 0) {
+			dev_err(gi2c->se.dev, "error turning off device :%d\n", ret);
+			return ret;
+		}
+	} else {
+		/* Increment the device usage counter and resume */
+		ret = pm_runtime_resume_and_get(gi2c->se.dev);
+		if (ret < 0) {
+			dev_err(gi2c->se.dev, "error turning on device :%d\n", ret);
+			return ret;
+		}
+		for (i = 0; i < CCU_MAX_GPII; i++) {
+			if (gi2c->tx_c[i])
+				dmaengine_alloc_chan_resources(gi2c->tx_c[i]);
+
+			if (gi2c->rx_c[i])
+				dmaengine_alloc_chan_resources(gi2c->rx_c[i]);
+		}
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL(geni_i2c_power_on_off);
+
+/*
+ * geni_i2c_send_cmd_gsi_mode - This function is used to send bus cmd
+ *			in GPI mode.
+ * @gi2c: Handle of the I2C device
+ * @cmd: opcode command
+ *
+ * Return: zero on success and error on failure.
+ */
+static int geni_i2c_send_cmd_gsi_mode(struct geni_i2c_dev *gi2c, int cmd)
+{
+	struct dma_slave_config config = {};
+	struct gpi_i2c_config peripheral = {};
+	struct dma_chan *dma_chan = gi2c->tx_c[0];
+	struct dma_async_tx_descriptor *desc;
+	dma_cookie_t cookie;
+	int ret;
+	unsigned int flags, time_left;
+	void *buf;
+	dma_addr_t addr;
+
+	config.peripheral_config = &peripheral;
+	config.peripheral_size = sizeof(peripheral);
+	peripheral.op = cmd;
+	peripheral.set_config = 1;
+	peripheral.multi_msg = false;
+
+	buf = kzalloc(1, GFP_KERNEL);
+	if (!buf)
+		return -ENOMEM;
+
+	addr = dma_map_single(gi2c->se.dev->parent, buf, 1, DMA_TO_DEVICE);
+	if (dma_mapping_error(gi2c->se.dev->parent, addr)) {
+		kfree(buf);
+		return -ENOMEM;
+	}
+
+	ret = dmaengine_slave_config(dma_chan, &config);
+	if (ret) {
+		dev_err(gi2c->se.dev, "dma config error: %d for op:%d\n", ret, cmd);
+		goto err_config;
+	}
+
+	peripheral.set_config = 0;
+
+	flags = DMA_PREP_INTERRUPT | DMA_CTRL_ACK;
+	desc = dmaengine_prep_slave_single(dma_chan, addr, 1, DMA_MEM_TO_DEV, flags);
+	if (!desc) {
+		dev_err(gi2c->se.dev, "prep_slave_sg failed\n");
+		ret = -EIO;
+		goto err_config;
+	}
+
+	desc->callback_result = i2c_gpi_cb_result;
+	desc->callback_param = gi2c;
+	cookie = dmaengine_submit(desc);
+	if (dma_submit_error(cookie)) {
+		dev_err(gi2c->se.dev, "%s: dmaengine_submit failed (%d)\n",
+			__func__, cookie);
+		goto err;
+	}
+
+	dma_async_issue_pending(dma_chan);
+	gi2c->err = 0;
+	reinit_completion(&gi2c->done);
+	time_left = wait_for_completion_timeout(&gi2c->done, XFER_TIMEOUT);
+	if (!time_left) {
+		dev_err(gi2c->se.dev, "I2C timeout for cmd:%d\n", cmd);
+		gi2c->err = -ETIMEDOUT;
+	}
+
+	if (gi2c->err) {
+		ret = gi2c->err;
+		goto err;
+	}
+
+	kfree(buf);
+	return 0;
+err:
+	dev_err(gi2c->se.dev, "GPI transfer failed: %d\n", ret);
+	dmaengine_terminate_sync(dma_chan);
+err_config:
+	dma_unmap_single(gi2c->se.dev->parent, addr, 1, DMA_TO_DEVICE);
+	kfree(buf);
+	return ret;
+}
+
+/*
+ *geni_i2c_send_cmd_fifo_mode - This function is used to send any command
+ *				in FIFO mode.
+ * @gi2c: Handle of the I2C device
+ * @cmd: Bus command
+ * Return: zero on success and error on failure.
+ */
+static int geni_i2c_send_cmd_fifo_mode(struct geni_i2c_dev *gi2c, int cmd)
+{
+	u32 m_param = 0;
+	u32 time_left = 0;
+
+	geni_se_select_mode(&gi2c->se, GENI_SE_FIFO);
+	reinit_completion(&gi2c->done);
+	geni_se_setup_m_cmd(&gi2c->se, cmd, m_param);
+	time_left = wait_for_completion_timeout(&gi2c->done, XFER_TIMEOUT);
+	if (!time_left) {
+		dev_err(gi2c->se.dev, "I2C timeout for cmd:%d\n", cmd);
+		return -ETIMEDOUT;
+	}
+
+	dev_dbg(gi2c->se.dev, "CMD:%d send successfully\n", cmd);
+	return 0;
+}
+
+/*
+ * geni_i2c_is_bus_recovery_required - Checks if Bus recovery required or not.
+ *
+ * @gi2c: Handle of the I2C device
+ *
+ * Return: TRUE if SDA is stuck LOW due to some issue else false.
+ */
+bool geni_i2c_is_bus_recovery_required(struct geni_i2c_dev *gi2c)
+{
+	u32 geni_ios = readl_relaxed(gi2c->se.base + SE_GENI_IOS);
+
+	/* SE_GENI_IOS will show I2C CLK/SDA line status, BIT 0 is SDA,
+	 * BIT 1 is CLK. SE_GENI_IOS register set when CLK/SDA line
+	 * is pulled high.
+	 */
+	return (((geni_ios & 1) == 0) && (gi2c->err == -EPROTO ||
+				 gi2c->err == -ETIMEDOUT));
+}
+
+/*
+ * geni_i2c_check_recover_ios_line - Check and recover IOS lines state.
+ *
+ * @adap: Master controller handle.
+ *
+ * Return: Success OR Respective error code/value.
+ */
+int geni_i2c_check_recover_ios_line(struct i2c_adapter *adap)
+{
+	struct geni_i2c_dev *gi2c;
+	int ret = 0;
+
+	if (!adap) {
+		pr_err("%s: Adap is NULL\n", __func__);
+		return -EINVAL;
+	} else {
+		gi2c = adap->dev.driver_data;
+		if (!gi2c) {
+			pr_err("%s: I2C driver data is NULL\n", __func__);
+			return -EINVAL;
+		}
+	}
+
+	if (geni_i2c_is_bus_recovery_required(gi2c)) {
+		dev_err(gi2c->se.dev, "IOS lines in bad state, need recovery\n");
+	} else {
+		dev_dbg(gi2c->se.dev, "IOS lines in good state\n");
+		return 0;
+	}
+	if (gi2c->gpi_mode) {
+		ret = geni_i2c_send_cmd_gsi_mode(gi2c, I2C_BUS_CLEAR);
+		if (ret) {
+			dev_err(gi2c->se.dev, "BUS CLEAR cmd failed for GPI mode\n");
+			return ret;
+		}
+
+		ret = geni_i2c_send_cmd_gsi_mode(gi2c, I2C_STOP_ON_BUS);
+		if (ret) {
+			dev_err(gi2c->se.dev, "STOP_ON_BUS cmd failed for GPI mode\n");
+			return ret;
+		}
+	} else {
+		ret = geni_i2c_send_cmd_fifo_mode(gi2c, I2C_BUS_CLEAR);
+		if (ret) {
+			dev_err(gi2c->se.dev, "BUS Clear cmd failed for FIFO mode\n");
+			return ret;
+		}
+
+		ret = geni_i2c_send_cmd_fifo_mode(gi2c, I2C_STOP_ON_BUS);
+		if (ret) {
+			dev_err(gi2c->se.dev, "STOP_ON_BUS cmd failed for FIFO mode\n");
+			return ret;
+		}
+	}
+
+	return ret;
+}
+EXPORT_SYMBOL(geni_i2c_check_recover_ios_line);
+
+/*
+ * geni_i2c_configure_ccu_mode : Configure the CCU mode and GPII channel type
+ *
+ * @adap: Master controller handle.
+ * @gpii_no: GPII number.
+ * @ch_type: Channel type for given GPII. (0 – Regular, 1 – sync ch 0, 2 – sync ch 1)
+ * @mode: Represents SE mode (0 – sync trigger mode and 1 – CDM mode)
+ *
+ * Return 0 on success and error on failure.
+ */
+int geni_i2c_configure_ccu_mode(struct i2c_adapter *adap, int gpii_no, int ch_type, bool mode)
+{
+	struct geni_i2c_dev *gi2c;
+	int idx = (gpii_no % CCU_MAX_GPII);
+
+	if (!adap) {
+		pr_err("%s: Adap is NULL\n", __func__);
+		return -EINVAL;
+	} else {
+		gi2c = adap->dev.driver_data;
+		if (!gi2c) {
+			pr_err("%s: I2C driver data is NULL\n", __func__);
+			return -EINVAL;
+		}
+	}
+
+	gi2c->ccu.mode = mode;
+	gi2c->ccu.gpi[idx].type = ch_type;
+	gi2c->ccu.gpi[idx].gpii_mode_change = true;
+
+	ccu_gpi_configure_mode(gi2c->tx_c[idx], mode, ch_type);
+	ccu_gpi_configure_mode(gi2c->rx_c[idx], mode, ch_type);
+
+	release_gpi_dma(gi2c);
+	setup_gpi_dma(gi2c);
+	gi2c->ccu.gpi[idx].gpii_mode_change = false;
+
+	return 0;
+}
+EXPORT_SYMBOL(geni_i2c_configure_ccu_mode);
+
+/*
+ * geni_i2c_gpii_reset : Reset the GPII channel.
+ *
+ * @adap: Master controller handle.
+ * @gpii_no: GPII number.
+ * @premature_cancel: client use this flag if want to drop the queued frames
+ *
+ * Return 0 on success and error on failure.
+ */
+int geni_i2c_gpii_reset(struct i2c_adapter *adap, int gpii_no,
+			bool premature_cancel)
+{
+	struct geni_i2c_dev *gi2c;
+	int idx = (gpii_no % CCU_MAX_GPII);
+	int ret = 0;
+
+	if (!adap) {
+		pr_err("%s: Adap is NULL\n", __func__);
+		return -EINVAL;
+	} else {
+		gi2c = adap->dev.driver_data;
+		if (!gi2c) {
+			pr_err("%s: I2C driver data is NULL\n", __func__);
+			return -EINVAL;
+		}
+	}
+
+	if (premature_cancel)
+		ccu_gpi_set_premature_cancel_flag(gi2c->tx_c[idx]);
+
+	if (gi2c->tx_c[idx]) {
+		ret = dmaengine_terminate_all(gi2c->tx_c[idx]);
+		if (ret) {
+			dev_err(gi2c->se.dev, "%s: gpi terminate failed for tx_c[%d]ret:%d\n",
+				__func__, ret, idx);
+			return ret;
+		}
+	}
+
+	if (gi2c->rx_c[idx]) {
+		ret = dmaengine_terminate_all(gi2c->rx_c[idx]);
+		if (ret) {
+			dev_err(gi2c->se.dev, "%s: gpi terminate failed for rx_c[%d]ret:%d\n",
+				__func__, ret, idx);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL(geni_i2c_gpii_reset);
+
+/*
+ * geni_i2c_configure_delay : Configure inter transfer delay.
+ *
+ * @adap: Master controller handle.
+ *
+ * @time : Delay in msec.
+ *
+ * Return none.
+ */
+void geni_i2c_configure_pre_delay(struct i2c_adapter *adap, unsigned long time)
+{
+	struct geni_i2c_dev *gi2c;
+
+	if (!adap) {
+		pr_err("%s: Adap is NULL\n", __func__);
+		return -EINVAL;
+	} else {
+		gi2c = adap->dev.driver_data;
+		if (!gi2c) {
+			pr_err("%s: I2C driver data is NULL\n", __func__);
+			return -EINVAL;
+		}
+	}
+
+	gi2c->ccu.inter_delay = time;
+}
+EXPORT_SYMBOL(geni_i2c_configure_pre_delay);
+
 static void geni_i2c_err_misc(struct geni_i2c_dev *gi2c)
 {
 	u32 m_cmd = readl_relaxed(gi2c->se.base + SE_GENI_M_CMD0);
@@ -655,7 +1029,8 @@ static void geni_i2c_gpi_unmap(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 
 static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 			struct dma_slave_config *config, dma_addr_t *dma_addr_p,
-			void **buf, unsigned int op, struct dma_chan *dma_chan)
+			void **buf, unsigned int op, struct dma_chan *dma_chan,
+			int idx)
 {
 	struct gpi_i2c_config *peripheral_cfg;
 	unsigned int flags;
@@ -664,6 +1039,7 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 	enum dma_data_direction map_dirn;
 	enum dma_transfer_direction dma_dirn;
 	struct dma_async_tx_descriptor *desc;
+	dma_cookie_t cookie;
 	int ret;
 
 	u16 len = gi2c->is_ccu_qup ? (msg->len - CCU_DATA_OFFSET) : msg->len;
@@ -697,8 +1073,22 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 		return -ENOMEM;
 	}
 
-	peripheral_cfg->rx_len = len;
+	if (op == I2C_WRITE && gi2c->is_ccu_qup) {
+		/* set the TX sub pkt_len */
+		peripheral_cfg->rx_len = gi2c->ccu.msg[idx]->pkt_len;
+		peripheral_cfg->ccu_cfg.rand_wr = true;
+	} else {
+		/* set the length as message for rx txn */
+		peripheral_cfg->rx_len = len;
+	}
 	peripheral_cfg->op = op;
+	peripheral_cfg->delay = gi2c->ccu.inter_delay;
+	peripheral_cfg->ccu_cfg.ccu_last = gi2c->ccu.msg[idx]->ccu_last;
+	peripheral_cfg->ccu_cfg.ch_last = gi2c->ccu.msg[idx]->ch_last;
+	peripheral_cfg->ccu_cfg.nop_last = gi2c->ccu.msg[idx]->nop_last;
+	peripheral_cfg->post_delay = gi2c->ccu.msg[idx]->post_delay;
+	peripheral_cfg->pre_delay = gi2c->ccu.msg[idx]->pre_delay;
+	peripheral_cfg->ccu_cfg.retry = gi2c->ccu.msg[idx]->arb_lost_retry;
 
 	ret = dmaengine_slave_config(dma_chan, config);
 	if (ret) {
@@ -709,6 +1099,14 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 
 	peripheral_cfg->set_config = 0;
 	peripheral_cfg->multi_msg = true;
+	peripheral_cfg->ccu_cfg.ccu_last = 0;
+	peripheral_cfg->ccu_cfg.ch_last = 0;
+	peripheral_cfg->ccu_cfg.nop_last = false;
+	peripheral_cfg->post_delay = false;
+	peripheral_cfg->pre_delay = false;
+	peripheral_cfg->ccu_cfg.retry = 0;
+	peripheral_cfg->rx_len = 0;
+	peripheral_cfg->ccu_cfg.rand_wr = false;
 
 	flags = DMA_PREP_INTERRUPT | DMA_CTRL_ACK;
 
@@ -728,7 +1126,13 @@ static int geni_i2c_gpi(struct geni_i2c_dev *gi2c, struct i2c_msg *msg,
 	desc->callback_result = i2c_gpi_cb_result;
 	desc->callback_param = peripheral_cfg;
 
-	dmaengine_submit(desc);
+	cookie = dmaengine_submit(desc);
+	if (dma_submit_error(cookie)) {
+		dev_err(gi2c->se.dev, "%s: dmaengine_submit failed (%d)\n",
+			__func__, cookie);
+		goto err_config;
+	}
+
 	*buf = dma_buf;
 	*dma_addr_p = addr;
 
@@ -867,19 +1271,35 @@ static int geni_i2c_gpi_xfer(struct geni_i2c_dev *gi2c, struct i2c_msg msgs[], i
 		peripheral->addr = msgs[i].addr;
 
 		ret =  geni_i2c_gpi(gi2c, &msgs[i], &config,
-				    &tx_addr, &tx_buf, I2C_WRITE, gi2c->tx_c[idx]);
+				    &tx_addr, &tx_buf, I2C_WRITE, gi2c->tx_c[idx], i);
 		if (ret)
 			goto err;
 
 		if (msgs[i].flags & I2C_M_RD) {
 			ret =  geni_i2c_gpi(gi2c, &msgs[i], &config,
-					    &rx_addr, &rx_buf, I2C_READ, gi2c->rx_c[idx]);
+					    &rx_addr, &rx_buf, I2C_READ, gi2c->rx_c[idx], i);
 			if (ret)
 				goto err;
 
 			dma_async_issue_pending(gi2c->rx_c[idx]);
 		}
 
+		/* Prioritize regular chan if Priority bit is set */
+		if (gi2c->is_ccu_qup && gi2c->ccu.msg[i]->priority) {
+			if (idx == REGULAR_GPII) {
+				dev_dbg(gi2c->se.dev, "Prioritize regular ch for msg[%d]\n", i);
+				ret = ccu_gpi_prioritize_chan(gi2c->tx_c[idx]);
+				if (ret) {
+					dev_err(gi2c->se.dev, "Failed to prioritize the msg[%d] ret:%d\n",
+						i, ret);
+					goto err;
+				}
+			} else {
+				dev_err(gi2c->se.dev, "Priority bit set for non regular ch. GPII:%d\n",
+					idx);
+			}
+		}
+
 		dma_async_issue_pending(gi2c->tx_c[idx]);
 
 		time_left = wait_for_completion_timeout(&gi2c->done, XFER_TIMEOUT);
@@ -1281,10 +1701,12 @@ static int geni_i2c_probe(struct platform_device *pdev)
 	init_completion(&gi2c->done);
 	spin_lock_init(&gi2c->lock);
 	platform_set_drvdata(pdev, gi2c);
-	ret = devm_request_irq(dev, gi2c->irq, geni_i2c_irq, IRQF_NO_AUTOEN,
-			       dev_name(dev), gi2c);
-	if (ret)
-		return dev_err_probe(dev, ret, "Request_irq failed: %d\n", gi2c->irq);
+	if (!gi2c->is_ccu_qup) {
+		ret = devm_request_irq(dev, gi2c->irq, geni_i2c_irq, IRQF_NO_AUTOEN,
+				       dev_name(dev), gi2c);
+		if (ret)
+			return dev_err_probe(dev, ret, "Request_irq failed: %d\n", gi2c->irq);
+	}
 
 	i2c_set_adapdata(&gi2c->adap, gi2c);
 	gi2c->adap.dev.parent = dev;
@@ -1350,7 +1772,8 @@ static int __maybe_unused geni_i2c_runtime_suspend(struct device *dev)
 	int ret = 0;
 	struct geni_i2c_dev *gi2c = dev_get_drvdata(dev);
 
-	disable_irq(gi2c->irq);
+	if (!gi2c->is_ccu_qup)
+		disable_irq(gi2c->irq);
 
 	if (gi2c->dev_data->geni_se_switch_state) {
 		ret = gi2c->dev_data->geni_se_switch_state(dev, false);
@@ -1377,7 +1800,8 @@ static int __maybe_unused geni_i2c_runtime_resume(struct device *dev)
 		}
 	}
 
-	enable_irq(gi2c->irq);
+	if (!gi2c->is_ccu_qup)
+		enable_irq(gi2c->irq);
 	gi2c->suspended = 0;
 	return ret;
 }
-- 
2.34.1

